File: ios/Reguerta/Reguerta/Presentation/ShiftCoverage/CoverageRehearsalViewModel.swift
import Foundation
import Observation

@MainActor
@Observable
final class CoverageRehearsalViewModel {
    let coverage: ShiftCoverageViewModel
    var email = ""
    var password = ""
    var casePath: [String] = []
    var draft: CoverageCommandDraft?
    private(set) var isSigningIn = false
    private(set) var loginFailed = false
    @ObservationIgnored private let access: any CoverageRehearsalAccess
    @ObservationIgnored private var draftSession: ShiftCoverageSession?
    @ObservationIgnored private var revision: UInt64 = 0

    init(access: any CoverageRehearsalAccess) {
        self.access = access
        coverage = ShiftCoverageViewModel(repository: access.repository)
    }

    var canSignIn: Bool { !isSigningIn && !email.trimmingCharacters(in: .whitespaces).isEmpty && !password.isEmpty }

    func signIn() async {
        guard canSignIn else { return }
        revision &+= 1
        let owner = revision
        let secret = password
        password = ""
        isSigningIn = true
        loginFailed = false
        do {
            let session = try await access.signIn(email: email.trimmingCharacters(in: .whitespaces), password: secret)
            guard revision == owner else { return }
            coverage.bind(session)
            await coverage.refresh()
        } catch {
            guard revision == owner else { return }
            loginFailed = true
        }
        guard revision == owner else { return }
        isSigningIn = false
    }

    func signOut() {
        revision &+= 1
        access.signOut()
        coverage.bind(nil)
        casePath = []
        password = ""
        draft = nil
        isSigningIn = false
        loginFailed = false
    }

    func openNotification(_ eventId: String) async {
        guard draft == nil else { return }
        casePath = []
        if let caseId = await coverage.openNotification(eventId) {
            casePath = [caseId]
        }
    }

    func reloadOverviewAfterNavigation() async {
        guard casePath.isEmpty, coverage.session != nil else { return }
        await coverage.refreshOverview()
    }

    var canOpen: Bool {
        !coverage.isBusy && coverage.pendingCommand == nil &&
            coverage.snapshot?.availableShifts?.contains(where: \.writable) == true
    }

    func present(_ action: ShiftCoverageCommand.Action, item: ShiftCoverageSnapshot.Case? = nil) {
        guard let snapshot = coverage.snapshot, !coverage.isBusy, coverage.pendingCommand == nil else { return }
        draftSession = coverage.session
        draft = CoverageCommandDraft(action: action, item: item, snapshot: snapshot, nowMillis: coverage.nowMillis)
    }

    var canConfirmDraft: Bool {
        guard let draft, let command = draft.command, let snapshot = coverage.snapshot,
              draftSession == coverage.session, !coverage.isBusy, coverage.pendingCommand == nil else { return false }
        if let expires = command.expiresAtMillis, expires <= coverage.nowMillis { return false }
        if command.action == .open {
            return snapshot.availableShifts?.contains {
                $0.shiftId == command.shiftId && $0.shiftRevision == command.expectedShiftRevision &&
                    $0.writable && $0.scheduledAtMillis > coverage.nowMillis
            } == true
        }
        guard let item = coverage.caseItem(command.caseId), item.revision == command.expectedRevision,
              item.shiftRevision == command.expectedShiftRevision else { return false }
        return coverage.actions(for: item).contains(command.action)
    }

    func confirmDraft() async {
        guard canConfirmDraft, let draft, let command = draft.command else { return }
        self.draft = nil
        await coverage.submit(command)
    }
}

File: android/Reguerta/app/src/main/java/com/reguerta/user/presentation/shiftcoverage/CoverageRehearsalViewModel.kt
package com.reguerta.user.presentation.shiftcoverage

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.reguerta.user.domain.shiftcoverage.CoverageRehearsalAccess
import com.reguerta.user.domain.shiftcoverage.ShiftCoverageCommand
import com.reguerta.user.domain.shiftcoverage.ShiftCoverageSession
import com.reguerta.user.domain.shiftcoverage.ShiftCoverageSnapshot
import kotlinx.coroutines.launch

internal class CoverageRehearsalViewModel(private val access: CoverageRehearsalAccess) : ViewModel() {
    val coverage = ShiftCoverageViewModel(access.repository)
    var selectedCaseId by mutableStateOf<String?>(null)
    var email by mutableStateOf("")
    var password by mutableStateOf("")
    var isSigningIn by mutableStateOf(false)
        private set
    var loginFailed by mutableStateOf(false)
        private set
    var draft by mutableStateOf<CoverageCommandDraft?>(null)
        private set
    private var draftSession: ShiftCoverageSession? = null
    private var revision = 0L
    val canSignIn: Boolean get() = !isSigningIn && email.isNotBlank() && password.isNotEmpty()

    fun signIn() {
        if (!canSignIn) return
        revision++
        val owner = revision
        val secret = password
        password = ""
        isSigningIn = true
        loginFailed = false
        viewModelScope.launch {
            try {
                val session = access.signIn(email.trim(), secret)
                if (revision != owner) return@launch
                coverage.bind(session)
                coverage.refresh()
            } catch (_: Exception) {
                if (revision == owner) loginFailed = true
            } finally {
                if (revision == owner) isSigningIn = false
            }
        }
    }

    fun signOut() {
        revision++
        access.signOut()
        coverage.bind(null)
        selectedCaseId = null
        password = ""
        draft = null
        isSigningIn = false
        loginFailed = false
    }

    val canOpen: Boolean get() = !coverage.state.value.isBusy && coverage.state.value.pendingCommand == null &&
        coverage.state.value.snapshot?.availableShifts?.any { it.writable } == true

    fun present(action: ShiftCoverageCommand.Action, item: ShiftCoverageSnapshot.Case? = null) {
        if (coverage.state.value.isBusy || coverage.state.value.pendingCommand != null) return
        val snapshot = coverage.state.value.snapshot ?: return
        draftSession = coverage.state.value.session
        draft = CoverageCommandDraft(action, item, snapshot, coverage.nowMillis)
    }

    fun dismissDraft() { draft = null }
    fun canConfirmDraft(now: Long = coverage.nowMillis): Boolean {
        val command = draft?.command ?: return false
        val state = coverage.state.value
        val snapshot = state.snapshot ?: return false
        if (draftSession != state.session || state.isBusy || state.pendingCommand != null) return false
        if (command.expiresAtMillis?.let { it <= now } == true) return false
        if (command.action == ShiftCoverageCommand.Action.open) {
            return snapshot.availableShifts.any { it.shiftId == command.shiftId &&
                it.shiftRevision == command.expectedShiftRevision && it.writable && it.scheduledAtMillis > now }
        }
        val item = snapshot.cases.firstOrNull { it.caseId == command.caseId } ?: return false
        return item.revision == command.expectedRevision && item.shiftRevision == command.expectedShiftRevision &&
            coverage.actions(item).contains(command.action)
    }

    fun confirmDraft() {
        if (!canConfirmDraft()) return
        val command = draft?.command ?: return
        draft = null
        viewModelScope.launch { coverage.submit(command) }
    }
    fun openNotification(eventId: String) {
        if (draft != null) return
        selectedCaseId = null
        viewModelScope.launch {
            coverage.openNotification(eventId)?.let { selectedCaseId = it }
        }
    }

    fun showOverview() {
        selectedCaseId = null
        viewModelScope.launch { coverage.refreshOverview() }
    }

    fun refresh() { viewModelScope.launch { coverage.refresh() } }
    fun retryPending() { viewModelScope.launch { coverage.retryPending() } }

    override fun onCleared() {
        signOut()
    }
}
